crates/driver/src/api.rs

```diff
@@ -1,38 +1,54 @@
 pub mod execute;
 pub mod solve;
 
+use crate::driver::Driver;
 use futures::Future;
 use shared::api::finalize_router;
-use std::net::SocketAddr;
+use std::{net::SocketAddr, sync::Arc};
 use tokio::{task, task::JoinHandle};
 use warp::{Filter, Rejection, Reply};
 
 pub fn serve_api(
     address: SocketAddr,
     shutdown_receiver: impl Future<Output = ()> + Send + 'static,
+    drivers: Vec<(Arc<Driver>, String)>,
 ) -> JoinHandle<()> {
-    let filter = handle_all_routes().boxed();
+    let filter = handle_all_routes(drivers).boxed();
     tracing::info!(%address, "serving driver");
     let (_, server) = warp::serve(filter).bind_with_graceful_shutdown(address, shutdown_receiver);
     task::spawn(server)
 }
 
-fn handle_all_routes() -> impl Filter<Extract = (impl Reply,), Error = Rejection> + Clone {
+fn handle_all_routes(
+    drivers: Vec<(Arc<Driver>, String)>,
+) -> impl Filter<Extract = (impl Reply,), Error = Rejection> + Clone {
     // Routes for api v1.
 
     // Note that we add a string with endpoint's name to all responses.
     // This string will be used later to report metrics.
     // It is not used to form the actual server response.
 
-    let solve = solve::post_solve().map(|result| (result, "solve")).boxed();
-    let execute = execute::post_execute()
-        .map(|result| (result, "execute"))
-        .boxed();
+    let mut base_routes = vec![];
+    for (driver, name) in drivers.into_iter() {
+        // leak string to use it in tracing spans
+        let name = Box::leak(name.into_boxed_str());
 
-    let routes = warp::path!("api" / ..)
-        .and(solve.or(execute).unify())
-        .untuple_one()
-        .boxed();
+        let solve = solve::post_solve(name, driver.clone())
+            .map(|result| (result, "solve"))
+            .boxed();
+        base_routes.push(solve);
```

Comment:
From the top of my head without thinking too deply: An alternative way to do this might be to not create a filter per solver but instead a single filter that handles all of these requests. That filter internally would extract the solver name from the url path and match it against the list of solvers it was configured with.

Reply:
Ah, neat idea - then you can just keep a `HashMap<String, Arc<Driver>>`.

Reply:
I'm a little hesitant about that idea. I would like to implement as little custom routing logic as possible. No need to reinvent the wheel.
If the current approach turns out to have some issues I have no problem with trying your approach but for now I would stick to the current one.

Reply:
I wouldn't necessarily call it custom routing logic. In our current route `/api/v1/account/{owner}/orders` we're not creating a new warp route for every possible owner. The owner is conceptually a parameter and all requests of this route get treated the same way. This is how I feel about the solver name. To me the path elements in a warp route that are not variables like `{owner}` should truly be static strings.
The observable result from the outside is the same so this is just a question of which implementation is cleaner. I am fine with either solution too so feel free to keep the current approach.